hex/helpers/lib/run.exs

```diff
@@ -40,37 +41,115 @@ defmodule DependencyHelper do
-    run_script("do_update.exs", dir, [dependency_name] ++ credentials)
+    set_credentials(credentials)
+
+    run_script("do_update.exs", dir, [dependency_name])
   end
 
   defp run_script(script, dir, args \\ []) do
-    args = [
-      "run",
-      "--no-deps-check",
-      "--no-start",
-      "--no-compile",
-      "--no-elixir-version-check",
-      script
-    ] ++ args
+    args =
+      [
+        "run",
+        "--no-deps-check",
+        "--no-start",
+        "--no-compile",
+        "--no-elixir-version-check",
+        script
+      ] ++ args
 
     System.cmd(
       "mix",
       args,
-      [
-        cd: dir,
-        env: %{
-          "MIX_EXS" => nil,
-          "MIX_LOCK" => nil,
-          "MIX_DEPS" => nil
-        }
-      ]
+      cd: dir,
+      env: %{
+        "MIX_EXS" => nil,
+        "MIX_LOCK" => nil,
+        "MIX_DEPS" => nil
+      }
     )
   end
+
+  defp set_credentials([]), do: :ok
+
+  defp set_credentials(["hex_organization", organization, token | tail]) do
+    url =
+      "hexpm"
+      |> Hex.Repo.get_repo()
+      |> Map.fetch!(:url)
+      |> URI.merge("/repos/#{organization}")
+      |> to_string()
+
+    update_repos("hexpm:#{organization}", %{url: url, public_key: nil, auth_key: token})
+
+    set_credentials(tail)
+  end
+
+  defp set_credentials(["hex_repo", repo, url, auth_key, fingerprint | tail]) do
+    case fetch_public_key(repo, url, auth_key, fingerprint) do
+      {:ok, public_key} ->
+        update_repos(repo, %{auth_key: auth_key, public_key: public_key, url: url})
+
+        set_credentials(tail)
+
+      error ->
+        handle_result(error)
+    end
+  end
+
+  defp set_credentials([_mode, org_or_url | _]) do
+    handle_result({:error, "Missing credentials for \"#{org_or_url}\""})
+  end
+
+  defp update_repos(name, opts) do
+    Hex.Config.read()
+    |> Hex.Config.read_repos()
+    |> Map.put(name, opts)
+    |> Hex.Config.update_repos()
+  end
+
+  defp fetch_public_key(repo, repo_url, auth_key, fingerprint) do
+    case Hex.Repo.get_public_key(repo_url, auth_key) do
+      {:ok, {200, key, _}} ->
+        if public_key_matches?(key, fingerprint) do
+          {:ok, key}
+        else
+          {:error, "Public key fingerprint mismatch for repo \"#{repo}\""}
+        end
+
+      {:ok, {code, _, _}} ->
+        {:error, "Downloading public key for repo \"#{repo}\" failed with code: #{inspect(code)}"}
+
+      other ->
+        {:error, "Downloading public key for repo \"#{repo}\" failed: #{inspect(other)}"}
+    end
+  end
+
+  defp public_key_matches?(_public_key, _fingerprint = ""), do: true
+
+  defp public_key_matches?(public_key, fingerprint) do
+    public_key =
+      public_key
+      |> :public_key.pem_decode()
+      |> List.first()
+      |> :public_key.pem_entry_decode()
+
+    decoded_fingerprint =
+      :sha256
+      |> :public_key.ssh_hostkey_fingerprint(public_key)
```

Comment:
This handles the fingerprint comparison using a built-in Erlang library. Note that it requires OTP 24 because that's when this function was added as part of a crypto overhaul.

Reply:
Is this different from [:ssh.hostkey_fingerprint/2](https://www.erlang.org/doc/man/ssh.html#hostkey_fingerprint-2)? 
```suggestion
      |> :ssh.hostkey_fingerprint(public_key)
```

Reply:
The difference appears to be from the OTP version. As of OTP 24 `hostkey_fingerprint/2` is available and documented, while `:ssh_hostkey_fingerprint` is no longer documented.

It's not apparent which Elixir/Erlang versions are used for CI or production. Feel free to apply this suggestion if you're on OTP 24+

Reply:
We do run OTP 24. I don't actually have an opinion on the implementation. I had initially encountered the docs when trying to understand the pull request. Since `hostkey_fingerprint/2` is documented, I'll apply the suggestion.